File: typescript/queue/queue.ts
type QueueNode<T> = {
    value: T;
    right: QueueNode<T> | null;
    left: QueueNode<T> | null;
};

export class Queue<T> {
    private head: QueueNode<T> | null;
    private tail: QueueNode<T> | null;
    private size: number;

    public constructor() {
        this.head = null;
        this.tail = null;
        this.size = 0;
    }

    private createNode(value: T): QueueNode<T> {
        return { value, left: null, right: null };
    }

    public enqueue(value: T): number {
        let node = this.createNode(value);
        if (this.size === 0) {
            this.head = node;
            this.tail = node;
        } else {
            this.tail!.left = node;
            node.right = this.tail;
            this.tail = node;
        }

        return this.size++;
    }

    public dequeue(): number {
        if (this.size === 0) return this.size;
        if (this.size === 1) {
            this.head = null;
            this.tail = null;
        } else {
            this.head!.left!.right = null;
            this.head = this.head!.left;
        }

        return this.size--;
    }

    public getSize() {
        return this.size;
    }

    public print() {
        let iterator = this.tail;
        while (iterator) {
            process.stdout.write(`[${iterator.value}] `);
            iterator = iterator.right;
        }
        process.stdout.write("\n");
    }

    public toString(): string {
        return `[TAIL: ${this.tail?.value}] <==> [HEAD: ${this.head?.value}]\nSize: ${this.size}`;
    }
}

File: c/queue/queue.c
#include "queue.h"
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

Node *create_node(const char *name) {
  Node *node = (Node *)malloc(sizeof(Node));

  node->name = name;
  node->prev = NULL;
  node->next = NULL;

  return node;
}

void enqueue(Queue *queue, const char *name) {
  Node *node = create_node(name);

  if (queue->size == 0) {
    queue->head = node;
  } else {
    queue->tail->prev = node;
    node->next = queue->tail;
  }

  queue->tail = node;
  queue->size++;
}

void dequeue(Queue *queue) {
  if (queue->size <= 0)
    return;
  Node *temp = queue->head;

  if (queue->size == 1) {
    queue->head = NULL;
    queue->tail = NULL;
  } else {
    queue->head->prev->next = NULL;
    queue->head = queue->head->prev;
  }

  queue->size--;
  free(temp);
}

void clear_queue(Queue *queue) {
  while (queue->head) {
    dequeue(queue);
  }
}

void print_queue(Queue queue) {
  Node *temp = queue.tail;

  while (temp) {
    printf("%s\n", temp->name);
    temp = temp->next;
  }
}

Queue *init_queue() {
  Queue *queue = (Queue *)malloc(sizeof(Queue));

  queue->head = NULL;
  queue->tail = NULL;
  queue->size = 0;

  return queue;
}

void destroy_queue(Queue* queue) {
  if (queue->size > 0) clear_queue(queue);
  free(queue);
}
